refactor: one webpack config for all sub-packages

taylrj opened the pull request:
- Kanban Flow Cards
  - [[orangutan] one webpack config for all sub-packages.](https://kanbanflow.com/t/6ketiGpC)
  - [[orangutan] bundle optimization](https://kanbanflow.com/t/6ZJTBqWu)
- The use cases this PR supports list as below if we want to include `scrollable-image` :
  - Use umd:  
    `<script src="https://unpkg.com/@twreporter/orangutan/dist/scrollable-image.js"> </script>`
  - Use esm:
    `import { ScrollableImage } from ‘@twreporter/orangutan`  or
    `import ScrollableImage from ‘@twreporter/orangutan/lib/scrollable-image/src`

taylrj commented on makefile:
```diff
@@ -9,13 +9,19 @@ dev: check-dep
+build-lib: check-dep
 	@echo "$(P) Run \`npm run build\` of all packages with \`scripts.build\`"
 	$(BIN_DIR)/lerna run --stream build
+	NODE_ENV=production $(BIN_DIR)/babel packages/index.js --out-dir lib --root-mode upward
```
We will have a `index.js` file (not included in this PR) like this:

```
    // index.js

    import ScrollableImage from './scrollable-image/src'

    export { ScrollableImage }
```

The reason to add this file is we need to support named imports, like:

```
   import { ScrollableImage } from '@twreporter/orangutan'
```

taylrj commented:
The folder structure after running `make build` would be

-- **dev**
-- **dist** 
   |-- *.js
-- **lib**
  |-- timeline/src 
  |-- scrollable-image/src
  |-- index.js
-- **packages** 
  |-- timeline
  |-- scrollable-image
  |-- index.js

nickhsine commented on webpack.config.js:
```diff
@@ -0,0 +1,134 @@
+    chunks: [],
+    bundles: [],
+  },
+}
```
Will we record our dependencies like the following pseudo code?
```
const webpackAssets = {
  scrollableImage: {
    chunks: [
      'react-$chunkhash-chunk.js',
      'twreporter-$chunkhash-chunk.js', 
      ... // omit due to explanation only
    ],
    bundles: [
      'scrollable-image-$chunkhash.bundle.js',
    ],
  }
}
```

Otherwise, how do we know what dependencies we need to load before executing?

taylrj replied:
Currently, webpackAssets are consumed in this way:

`
  const packageName = 'scrollable-image'
  const assets = [ ...chunks, ...bundles ]
  const assetScript = assets.map(src => {
      if (src.endsWith('bundle.js')) {
        if (src.indexOf(`${packageName}`) !== -1) {
          return `<script type="text/javascript" src="${src}"></script>`
        }
        return
      }
      return `<script type="text/javascript" src="${src}"></script>`
    })
`

taylrj replied:
The reason of this design is our packages would have common chunks, so the only difference is name of `bundle.js`. Currently, I use the name provided in entry point of webpack config file as the packageName for picking the desired script.

nickhsine replied:
Named bundle files are ok for loading.
However, in the above example,
will we load all chunks even if `scrollable-image` does not need it?

nickhsine replied:
Since the future repo will contain three packages,
`scrollable-image`, `timeline` and `dual-channel`.

`dual-channel` might generate different chunks which `scrollable-image` doesn't need.
In current `webpack-asset.json`, how to only load necessary chunks for each sub-package?

taylrj replied:
Updated. Please see commit `b959092` and `2715158` for explanation. ( draft #4 is also updated.)

In these patches, the common chunks are in dist/ folder and packages' own chunks would be in their folder (e.g. dist/scrollable-image).

Here's folder structure:

|- dist/
|--dual-channel/
|----dual-channel-chunk.js
|----dual-channel-main.js
|--scrollable-image/
|----scrollable-image-main.js
|--timeline/
|--common-1-chunk.js
|--common-2-chunk.js

taylrj commented:
@nickhsine 

Please look at commit [6cf013a](https://github.com/twreporter/orangutan/pull/4/commits/6cf013a96e2383298821a687a36a7b5bc04f7c38) in draft #4  only.

Files added in #4 can build dist/ and lib/ files.

taylrj commented:
@nickhsine 

Updated.

Haven't found proper hook yet.

I updated `webpackAssets` structure to meet our requirement in [this commit](https://github.com/twreporter/orangutan/pull/3/commits/db9f986e3f1ca845bff2fc871a53391312c53326).

Please have a see.

nickhsine commented:
LGTM  👍 💯 🥇
